nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/CommentsMigrator.java

```diff
@@ -59,21 +73,67 @@
 
     protected static final int BATCH_SIZE = 50;
 
+    /**
+     * /**
+     *
+     * @since 11.1.
+     */
+    public final static String GET_HIDDEN_FOLDERS_QUERY = String.format("SELECT %s FROM Document WHERE %s='%s'",
+            ECM_UUID, ECM_PRIMARYTYPE, HIDDEN_FOLDER_TYPE);
+
     @Override
     protected String probeSession(CoreSession session) {
-        CommentService commentComponent =  Framework.getService(CommentService.class);
+        CommentService commentComponent = Framework.getService(CommentService.class);
         CommentServiceConfig commentServiceConfig = commentComponent.getConfig();
         if (commentServiceConfig != null) {
             Graph graph = Framework.getService(RelationManager.class).getGraph(commentServiceConfig.graphName, session);
-            if (graph.getStatements().size() > 0) {
+            if (!graph.getStatements().isEmpty()) {
                 return MIGRATION_STATE_RELATION;
             }
         }
-        return MIGRATION_STATE_PROPERTY;
+        // If not in relation, check if there are still property comments under hidden Comments folder(s)
+        // Do a new query one to check the step and the second the make the migration (do all comments)
+        if (hasUnsecuredComments(session)) {
+            return MIGRATION_STATE_PROPERTY;
+        }
+        // Else, comments are already secured
+        return MIGRATION_STATE_SECURED;
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> getUnsecuredComments(CoreSession session) {
+        return getCommentFolders(session).stream()
+                                         .flatMap(f -> session.getChildren(f.getRef(), COMMENT_DOC_TYPE).stream())
+                                         .collect(Collectors.toList());
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected boolean hasUnsecuredComments(CoreSession session) {
+        List<DocumentModel> folders = getCommentFolders(session);
+        return folders.stream().anyMatch((folder) -> session.hasChildren(folder.getRef()));
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> getCommentFolders(CoreSession session) {
+        return session.query(GET_HIDDEN_FOLDERS_QUERY, (doc) -> doc.getName().equals(COMMENTS_DIRECTORY)
+                || session.getDocument(doc.getParentRef()).getName().equals(COMMENTS_DIRECTORY));
     }
 
     @Override
     protected void migrateSession(CoreSession session) {
```

Comment:
This API shouldn't exist in migrators, see my previous comments.

Reply:
We end up by using a NXQL with the `IN` operator : 
- To get all comments folders `Ids` (Comments under Domain + the Sub folders under these Comments) using session#queryProjection
- To get all comments documents, unfortunately we need the comment  document model to move it from the current path to the new one, 
` session.move(comment.getRef(), commentManager.getAncestorRef(session, comment), null);`  i don't know if we can get the AncestorIds (comment:ancestorIds) in projection our idea is to replace the call on `commentManager.getAncestorRef` and if not what is the best way to get this document model directly by using `session#query` or by getting the document unitary `session#getDocument` (then it will be use the session cache if its loaded before, but we can end up with memory issu if there is many comments)